Mk verify translation upload

mkangia opened the pull request:
[Trello Card](https://trello.com/c/jrXUocJu/218-work-on-ways-to-automate-the-currently-manual-processes-involved-with-excel-checking-an-upload-that-still-remain-post-transifex)

This adds the ability for a user to validate app translations before updating the app. Currently behind a FF.

Adds a checkbox for user to just validate translations and share issues if found over email.

<img width="514" alt="screen shot 2018-06-25 at 12 22 12 am" src="https://user-images.githubusercontent.com/3864163/41822488-fd30461c-780d-11e8-9442-bcb3c64f1ba5.png">


Issues found are shared over email which would look like this

<img width="595" alt="screen shot 2018-06-18 at 5 08 33 pm" src="https://user-images.githubusercontent.com/3864163/41534043-b2770cf2-731a-11e8-9ebf-0bec6559f057.png">

For review,
can look at first 2 commits for the major changes, after that it has been a series of improvements so can review all at once.

mkangia committed:
extract read_uploaded_app_translation_file

mkangia committed:
add validator

mkangia committed:
add changes to view

mkangia committed:
update form to set to verify

mkangia committed:
reorder fields to make form customization less ugly

mkangia committed:
read uploaded file at a single place and proactively

mkangia committed:
extract COLUMNS_TO_COMPARE and minor improvs

mkangia committed:
remove redundant arg expected_rows and default_language

mkangia committed:
add a new class for form instead

mkangia committed:
verify -> validate

mkangia committed:
remove individual methods for each type

mkangia committed:
fix poor implementation

append caused the original list to be updated each time this was done

mkangia committed:
extract message

mkangia committed:
fix concat error

mkangia committed:
add ability to share discrepancies over email

mkangia committed:
better message

mkangia committed:
lint

kaapstorm commented on corehq/apps/app_manager/app_translations/validator.py:
```diff
@@ -0,0 +1,88 @@
+            iterate_on = [expected_rows, uploaded_rows]
+        else:
+            iterate_on = [expected_rows, uploaded_rows]
+        for i, (expected_row, uploaded_row) in enumerate(zip(*iterate_on)):
```
If number_of_uploaded_rows < number_of_expected_rows then `uploaded_rows` and `expected_rows` will get switched around, because on line 55 you're putting `uploaded_rows` first, and here you're pulling them out as `expected_rows`.

You can set `iterate_on` outside of the `if` clause because `zip()` will stop after it reaches the end of the shortest list, regardless of whether it's the first or second argument.

kaapstorm commented:
(Test failure looks related. (Sorry for interrupted code review. I'll be able to get back to this in the morning.))

mkangia committed:
fix issue with switching order

switching the order would make the later iteration incorrect where it expect the first value to be the expected value
plus its not needed anyway

mkangia committed:
fix tests

kaapstorm commented on corehq/apps/app_manager/app_translations/app_translations.py:
```diff
@@ -49,7 +52,26 @@ def none_or_unicode(val):
+            ).format(e))
+        )
+        return False
+    return workbook
```
Not a criticism, just curious if there is an underlying principle behind your choice here: You've chosen to return `workbook`, but pass `msgs` in by reference, and update it as a side-effect, but not read from it. The more obvious would be ...
```python
def read_uploaded_app_translation_file(f):
    msgs = []
    # ... etc. ...
    return workbook, msgs
```
... so I think you probably considered that and intentionally chose the approach you took. 

I'm not suggesting you change your code to this. I just want to learn why sometimes one would want to return some values explicitly, and other values by reference (even when you don't want to read the incoming value).

mkangia replied:
So in this case it would be 
```
def read_uploaded_app_translation_file(f):
    msgs = []
    # ... etc. ...
    return workbook, msgs
```
and then where this is called
```
    workbook, msgs = read_uploaded_app_translation_file(f)
```

the only advantage in the current implementation i see is that we have `msgs` declared only once whereas in the code snippet above we have it at two places and there is no need for them to keep separate variables. 

I guess the current implementation would make more sense if `msgs` was not an empty string but with some messages already added earlier before its passed to `read_uploaded_app_translation_file` and we would then modify the same if needed.
If the change makes things more clear, happy to change it. Passing by reference has confused me at times as well but i guess that is just a habit thing.

kaapstorm replied:
Yeah, I agree that changing values passed by reference can be confusing, and harder to see what's happening. (But sometimes it's very useful.)

I think what you're doing here is obvious and not confusing at all, so happy happy to keep this.

mkangia replied:
updated to return msgs from this method.

mkangia commented on corehq/apps/hqwebapp/forms.py:
```diff
@@ -99,6 +93,33 @@ def __init__(self, plural_noun, action, form_id, *args, **kwargs):
+        ]
+
+
+class AppTranslationsBulkUploadForm(BulkUploadForm):
```
@orangejenny do you have suggestion on doing this better?
i had to add two fields to BulkUploadForm which is used at multiple places. My first approach was [this](https://github.com/dimagi/commcare-hq/pull/20939/commits/ef459ccfe7748bc406775a1425dd860f779390cc#diff-d39e56548bf4e0c07edcb0998f4aef34). But then i decided to make it less complicated by adding a new class instead of modifying `BulkUploadForm`.

orangejenny replied:
This seems all right; I like subclassing better than the previous code that added logic to `BulkUploadForm`. This class should into corehq/apps/app_manager/forms.py

mkangia replied:
haha.. I found editing `BulkUploadForm` really ugly @esoergel .. by adding this subclass it helps extract out the logic required only for `AppTranslationsBulkUploadForm` separately so don't think one editing `BulkUploadForm` would break anything here.

orangejenny commented:
What would you think of changing the UI, so instead of having the two checkboxes, there was just a second button next to "Upload App Translations" that said something like "Validate" with an envelope icon, and it always emails? I'd like to remove one of the options to reduce decision fatigue, and the email looks better than the in-page errors. Button instead of checkbox because I think validating is conceptually a different action than uploading, with very different results.

orangejenny commented:
Limiting this to email-only would also mean you could do the validation asynchronously, which might be nice - how long does the validation take for an ICDS-sized app?

orangejenny commented on corehq/apps/app_manager/app_translations/app_translations.py:
```diff
@@ -183,6 +189,40 @@ def process_bulk_app_translation_upload(app, f):
+
+
+@task(queue="email_queue")
+def _email_app_translations_discrepancies(msgs, email, app_name):
```
I'd use `send_html_email_async` instead; it has retry logic built in and deals with both a text and html version of the email.

orangejenny commented on corehq/apps/app_manager/app_translations/app_translations.py:
```diff
@@ -183,6 +189,40 @@ def process_bulk_app_translation_upload(app, f):
+
+@task(queue="email_queue")
+def _email_app_translations_discrepancies(msgs, email, app_name):
+    message = '<br/><br/>'.join([
```
You could build the string with newlines, which would read a bit more easily, and then use `linebreaksbr` to make it look reasonable as HTML.

orangejenny commented on corehq/apps/app_manager/app_translations/validator.py:
```diff
@@ -0,0 +1,86 @@
+    'module': ['case_property', 'list_or_detail'],
+    'form': ['label'],
+}
+ROW_COUNT_MISMATCH_MESSAGE = "We found {} rows than expected. Validating what was uploaded."
```
translate?

mkangia replied:
@esoergel more/less is actually added later on depending upon the count.

> translate?
Looks like i cant add translation to it if i add more/less later. Going to break it down to two different message instead.

esoergel replied:
Ah, I thought the `{}` was going to be replaced with a number.  Makes sense.

orangejenny replied:
There's also `We found {} rows but expected {}` as an alternative.

orangejenny commented on corehq/apps/app_manager/app_translations/validator.py:
```diff
@@ -0,0 +1,86 @@
+ROW_COUNT_MISMATCH_MESSAGE = "We found {} rows than expected. Validating what was uploaded."
+
+
+class UploadedTranslationsValidator(object):
```
Are these new validations? Are they errors that currently break bulk translations but we don't surface them to users?

mkangia replied:
> Are they errors that currently break bulk translations

I think some of them would still get fetched for example [here](https://github.com/dimagi/commcare-hq/blob/1b6bdda8bd5c4dad18f8abc0ce99e3ad010991f2/corehq/apps/app_manager/app_translations/app_translations.py#L914) but then the rest of the update would happen in the current process. Down the road we can probably combine these two implementations in a way.. but this one is just to automate the manual process one would follow to compare two excel sheets(expected vs to be uploaded) and share possible discrepancies.

orangejenny replied:
> Down the road we can probably combine these two implementations in a way

This is the kind of thing that doesn't tend to happen for a while, if ever, and gets harder to do the more time passes. It seems unfortunate that you've spent the time on this feature but that it'll only benefit ICDS, when bulk translation errors are a known pain point for other projects.

orangejenny commented on corehq/apps/app_manager/views/translations.py:
```diff
@@ -99,13 +103,21 @@ def download_bulk_app_translations(request, domain, app_id):
+    msgs = []
+    workbook = read_uploaded_app_translation_file(request.file, msgs)
+    if workbook:
+        if validate:
```
The validation code should also be run when the actual upload happens, either here or inside process_bulk_app_translation_upload.

mkangia replied:
hmm, so the way i see is that one would upload a file and check if its good and then simply translate with that same file if all went well. Checking again would ensure if the app has changed in between those two actions..

orangejenny replied:
Yeah, I agree that's the most likely workflow, but if validation is quick, I think doing it again is worthwhile. We double up on validation relatively often - e.g., some validation is done both client-side and server-side - for the sake of making HQ more robust.

orangejenny reviewed (changes requested):
Left a couple comments/questions.

esoergel commented on corehq/apps/app_manager/app_translations/validator.py:
```diff
@@ -0,0 +1,86 @@
+                uploaded_value = uploaded_row.get(column_name)
+                expected_value = expected_row[self._get_header_index(sheet_name, column_name)]
+                if expected_value != uploaded_value:
+                    # i + 2 to account for the sheet header as well
```
`enumerate` takes an optional second arg that's the index to start on, so you can do:
```python
>>> list(enumerate('abcde', 2))
[(2, 'a'), (3, 'b'), (4, 'c'), (5, 'd'), (6, 'e')]
```

esoergel commented on corehq/apps/app_manager/app_translations/validator.py:
```diff
@@ -0,0 +1,86 @@
+        """
+        columns_to_compare = COLUMNS_TO_COMPARE[for_type] + [self.default_language_column]
+        expected_rows = self.expected_rows[sheet_name]
+        msg = []
```
nit: I'd expect `msg` to refer to a single message, rather than a list of messages.

mkangia replied:
replaced with `msgs`

mkangia commented on corehq/apps/hqwebapp/forms.py:
```diff
@@ -99,6 +93,33 @@ def __init__(self, plural_noun, action, form_id, *args, **kwargs):
+                'bulk_upload_file',
+                data_bind="value: file",
+            )
+        ]
```
oh that was a miss.. but since the action field is a hidden field, didnt see any issue in UI. Will fix

mkangia committed:
review fixes

- use sent_html_email_async
- better message creation by using \n and linebreaksbr
- msg -> msgs
- pass 2 to enumerate for index jump
- use string_to_boolean
- revert change in order of fields

mkangia committed:
add translation to messages

mkangia committed:
always send email in case of errors

mkangia commented:
@orangejenny 

> What would you think of changing the UI, so instead of having the two checkboxes, there was just a second button next to "Upload App Translations" that said something like "Validate" with an envelope icon, and it always emails?

i think that would be a better UI for sure.. though i don't feel positive about adding two buttons to a form. Plus the code change for this would be to allow subclasses for `BulkUploadForm` to add buttons and also would need to find a way to align them next to one another. Not sure if this would be worth the code changes..

mkangia commented:
@orangejenny 
> Limiting this to email-only would also mean you could do the validation asynchronously, which might be nice - how long does the validation take for an ICDS-sized app?

Limiting this to email-only. I have not tested this for ICDS app but i don't expect this to take long to validate. Plus one would want to validate it and get the result of it before performing the validation so i don't think making it async would add benefit for the user.

mkangia committed:
choose better readability by avoiding passing msgs

mkangia commented on corehq/apps/app_manager/views/translations.py:
```diff
@@ -99,13 +103,21 @@ def download_bulk_app_translations(request, domain, app_id):
 @require_can_edit_apps
 @get_file("bulk_upload_file")
 def upload_bulk_app_translations(request, domain, app_id):
+    validate = request.POST.get('validate')
```
i think its fine. `upload` sees what action is required and then either `process` it to update the translations or `validate` it. I think you are taking `upload` as upload translations and i am reading this as `upload` translations file.

mkangia replied:
> You could use string_to_boolean here

Had some trouble with this. the value for `validate` is `on` if the checkbox is checked and is blank/`None` otherwise. Made the change to use `string_to_boolean` but then reverted it.

esoergel replied:
Oh, that makes sense.  I think `string_to_boolean` is meant for interpreting user-specified strings as booleans.  If this comes from something automatic, then it'd probably make more sense to check for an exact value.

orangejenny replied:
> This parameter changes the meaning of the view

FWIW I agree with Ethan here, it'd be better to do two views (that would likely share code).

mkangia committed:
revert string_to_boolean

esoergel commented on corehq/apps/app_manager/app_translations/validator.py:
```diff
@@ -7,13 +7,16 @@
 }
-ROW_COUNT_MISMATCH_MESSAGE = "We found {} rows than expected. Validating what was uploaded."
+LESS_ROW_COUNT_MISMATCH_MESSAGE = _("We found less rows than expected. Validating what was uploaded.")
+MORE_ROW_COUNT_MISMATCH_MESSAGE = _("We found more rows than expected. Validating what was uploaded.")
```
I don't think this'll work, since this is loaded at import time, and there isn't yet a translation context. Might have to use `ugettext_lazy` or define these in the body of the function somewhere.

mkangia replied:
thanks, using `ugettext_lazy`

mkangia committed:
use ugettext_lazy

mkangia committed:
lint

orangejenny commented:
About the proposed UI change: Two buttons is fine as long the primary one (uploading) is visually more prominent. But I didn't realize that would mean fiddling more with the bulk upload code, agree it's not worth it for this feature.

orangejenny reviewed (approved):
I still think there's some feedback worth considering, but nothing blocking.
